third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -0,0 +1,286 @@
+  initOpInsertion(lLoadOps[0]);
+  attachOp(builder.create<ROCDL::SetPrioOp>(loc, highPrioAttr));
+  attachOp(gLoadOps[0]);
+  attachOp(builder.create<ROCDL::SchedBarrier>(loc, 0));
+  attachOp(lLoadOps[1]);
+  attachOp(builder.create<ROCDL::SetPrioOp>(loc, lowPrioAttr));
+  attachOp(gLoadOps[1]);
+  attachOpWithPrio(builder, dotOps[0], loc);
+}
+
+void Pingponger::genOffsetConstants(Location loc, OpBuilder &builder,
+                                    unsigned numSlices, int64_t sliceWidth) {
+  for (int i = 0; i < numSlices; i++) {
+    int64_t offset = sliceWidth * i;
+    constOffsets.push_back(
+        builder.create<arith::ConstantIntOp>(loc, offset, 32));
+  }
+}
+
+LogicalResult Pingponger::genLocalSlice(OpBuilder &builder, Value v,
+                                        Attribute dotEncoding, unsigned opIdx,
+                                        unsigned numSlices,
+                                        int64_t sliceWidth) {
+  SmallVector<Operation *> slices;
+  SmallVector<Operation *> subviews;
+  auto memDesc = v.getDefiningOp()->getOperand(0);
+  auto type = cast<ttg::MemDescType>(memDesc.getType());
+  auto encoding = cast<RankedTensorType>(v.getType()).getEncoding();
+  auto srcEncoding = mlir::cast<ttg::DotOperandEncodingAttr>(encoding);
+  SmallVector<int64_t> shape{type.getShape().begin(), type.getShape().end()};
+  Type elementType = type.getElementType();
+  int64_t kIdx = opIdx == 0 ? 1 : 0;
+  shape[kIdx] = sliceWidth;
+  if (sliceWidth < 16)
+    return failure();
+
+  for (int i = 0; i < numSlices; i++) {
+    SmallVector<Value> offsetsVal;
+    SmallVector<int64_t> offsets = {0, 0};
+    offsets[kIdx] = i;
+    for (int64_t off : offsets) {
+      offsetsVal.push_back(constOffsets[off]);
+    }
+
+    Value newSmem = builder.create<triton::gpu::MemDescSubviewOp>(
+        v.getLoc(),
+        ttg::MemDescType::get(shape, elementType, type.getEncoding(),
+                              type.getMemorySpace()),
+        memDesc, offsetsVal);
+    auto dotOperandEnc = triton::gpu::DotOperandEncodingAttr::get(
+        builder.getContext(), opIdx, dotEncoding, srcEncoding.getKWidth());
+    Value prefetchSlice = builder.create<triton::gpu::LocalLoadOp>(
+        v.getLoc(), RankedTensorType::get(shape, elementType, dotOperandEnc),
+        newSmem);
+    subviews.push_back(newSmem.getDefiningOp());
+    slices.push_back(prefetchSlice.getDefiningOp());
+  }
+
+  subViewOps.push_back(subviews);
+  loadSliceOps.push_back(slices);
+  return success();
+}
+
+// Transform two warps coming from the same blocks
+LogicalResult Pingponger::transformTwoWarp(OpBuilder &builder, Location loc) {
+  // First, slice local_loads and dot into 4 parts
+  unsigned numSlices = 4;
+  auto op = cast<triton::DotOp>(dotOps[0]);
+  builder.setInsertionPointToStart(forOp.getBody());
+  auto typeB = op.getB().getType();
+  auto shapeB = typeB.getShape();
+  int64_t sliceWidth = shapeB[0] / numSlices;
```

Comment:
This is why it checks slice width can't < 16,  later mfma lowering should be always fine with it.
I'm struggling to find better way because mfma instruction is not selected in this stage.

Reply:
I meant just check `shapeB[0] % numSlices == 0` to be safe. (it's likely true but you never know what bugs we can have..)

Reply:
right, done.